DYDX-Bots/Approve_Spending_Monitor/src/agent.spec.ts

```diff
@@ -0,0 +1,223 @@
+import { BigNumber } from "ethers";
+import { Interface } from "ethers/lib/utils";
+import { FindingType, FindingSeverity, Finding, HandleTransaction, TransactionEvent } from "forta-agent";
+import { MockEthersProvider } from "forta-agent-tools/lib/mock.utils";
+import { createAddress, TestTransactionEvent } from "forta-agent-tools/lib/tests.utils";
+import { provideHandleTransaction } from "./agent";
+import BalanceFetcher from "./balance.fetcher";
+import { APPROVAL_EVENT, BALANCEOF_ABI } from "./utils";
+
+describe("Large spending approval tests suite", () => {
+  let handler: HandleTransaction;
+  let mockProvider = new MockEthersProvider();
+  const EVENT_IFACE = new Interface(APPROVAL_EVENT);
+  const mockNetworkManager = {
+    dydxAddress: createAddress("0xa1"),
+    usdcAddress: createAddress("0xa2"),
+    safetyModule: createAddress("0xa3"),
+    liquidityModule: createAddress("0xa4"),
+  };
+  const mockFetcher = new BalanceFetcher(mockProvider as any, mockNetworkManager as any);
+  const TEST_PERCENTAGE = 20;
+  const TOTAL_STAKED = 1000;
+
+  const TEST_DATA: [string, string, BigNumber][] = [
+    [createAddress("0xb01"), createAddress("0xb02"), BigNumber.from(300)], // above threshold
+    [createAddress("0xb11"), createAddress("0xb21"), BigNumber.from(500)], // above threshold
+    [createAddress("0xb12"), createAddress("0xb22"), BigNumber.from(20)], // below threshold
+    [createAddress("0xb13"), createAddress("0xb23"), BigNumber.from(100)], // below threshold
+  ];
+  const TEST_BLOCKS = [10, 20, 30, 40];
+  const createBalanceOfcall = (tokenAddress: string, module: string, balance: BigNumber, blockNumber: number) => {
+    mockProvider.addCallTo(tokenAddress, blockNumber, new Interface(BALANCEOF_ABI), "balanceOf", {
```

Comment:
Could probably create a `interface` variable rather than having `new Interface` every time `createBalanceOfCall` is called.

Same thing with `balance.fetcher.spec.ts`